source/extensions/common/dynamic_forward_proxy/dns_cache_resource_manager.h

```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+#include <atomic>
+#include <memory>
+#include <string>
+
+#include "envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.pb.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats_macros.h"
+#include "envoy/upstream/resource_manager.h"
+
+#include "common/common/assert.h"
+#include "common/common/basic_resource_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace DynamicForwardProxy {
+
+#define ALL_DNS_CACHE_CIRCUIT_BREAKERS_STATS(OPEN_GAUGE, REMAINING_GAUGE)                          \
+  OPEN_GAUGE(rq_pending_opening, Accumulate)                                                       \
+  REMAINING_GAUGE(rq_pending_remaining, Accumulate)
+
+struct DnsCacheCircuitBreakersStats {
+  ALL_DNS_CACHE_CIRCUIT_BREAKERS_STATS(GENERATE_GAUGE_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
+class DnsCacheResourceImpl : public BasicResourceLimitImpl {
+public:
+  using Base = BasicResourceLimitImpl;
+  explicit DnsCacheResourceImpl(uint64_t max, Runtime::Loader& runtime,
+                                const std::string& runtime_key, Stats::Gauge& opening,
+                                Stats::Gauge& remaining)
+      : Envoy::BasicResourceLimitImpl(max, runtime, runtime_key), opening_(opening),
+        remaining_(remaining) {
+    remaining_.set(max);
+  }
+
+  void inc() override {
+    Base::inc();
+    remaining_.set(Base::canCreate() ? Base::max() - current_ : 0);
+    opening_.set(Base::canCreate() ? 0 : 1);
+  }
+
+  void decBy(uint64_t amount) override {
+    Base::decBy(amount);
+    remaining_.set(Base::canCreate() ? Base::max() - current_ : 0);
+    opening_.set(Base::canCreate() ? 0 : 1);
+  }
+
+private:
+  Stats::Gauge& opening_;
+  Stats::Gauge& remaining_;
```

Comment:
It seems like we do this pattern elsewhere? Can this be generalized also somewhere? cc @tonya11en